chainer/utils/type_check.py

```diff
@@ -0,0 +1,281 @@
+import numpy
+
+from chainer import cuda
+
+
+class TypeInfo(object):
+
+    def __init__(self, name, index, shape, dtype):
+        self.index = index
+        name = '{0}[{1}]'.format(name, index)
+        self.name = name
+        self.shape = tuple(Shape(x, i, name) for i, x in enumerate(shape))
+        self.dtype = DtypeExpr(dtype, name)
+        self.ndim = Member(len(self.shape), name, 'ndim')
+
+    def is_none(self):
+        return self.dtype is None
+
+    def __str__(self):
+        return self.name
+
+
+class TypeInfoTuple(tuple):
+
+    def _set_name(self, name):
+        self.name = name
+
+    def size(self):
+        return Member(len(self), self.name, 'size')
+
+
+def get_types(data, is_grad):
+    assert(isinstance(data, tuple))
+
+    if is_grad:
+        name = 'out_types'
+    else:
+        name = 'in_types'
+
+    info = TypeInfoTuple(
+        get_type(name, i, x, is_grad) for i, x in enumerate(data))
+    info._set_name(name)
+    return info
+
+
+def get_type(name, index, array, accept_none):
+    if accept_none and array is None:
```

Comment:
If `get_type` is not called only from get_types, I think `get_type` should be in `get_types` to narrow its scope.